core/thread/inc/ROOT/TThreadedObject.hxx

```diff
@@ -173,84 +161,112 @@ namespace ROOT {
-         Create(nslots, args...);
+         // create at least one directory (we need it for fModel), plus others as needed by the size of fObjPointers
+         fDirectories.emplace_back(Internal::TThreadedObjectUtils::DirCreator<T>::Create());
+         for (auto i = 1u; i < nSlots; ++i)
+            fDirectories.emplace_back(Internal::TThreadedObjectUtils::DirCreator<T>::Create());
+
+         TDirectory::TContext ctxt(fDirectories[0]);
+         fModel.reset(Internal::TThreadedObjectUtils::Detacher<T>::Detach(new T(std::forward<ARGS>(args)...)));
       }
 
-      /// Construct the TThreaded object and the "model" of the thread private
-      /// objects, assuming IMT usage. This constructor is equivalent to `TThreadedObject(kIMTPoolSize,...)`
+      /// Construct the TThreadedObject and the "model" of the thread private objects.
       /// \tparam ARGS Arguments of the constructor of T
       template<class ...ARGS>
-      TThreadedObject(ARGS&&... args) : fIsMerged(false)
+      TThreadedObject(ARGS&&... args) : TThreadedObject(TNumSlots{fgMaxSlots}, args...) { }
+
+      /// Return the number of currently available slot.
+      ///
+      /// The method is safe to call concurrently to other TThreadedObject methods.
+      /// Note that slots could be available but contain no data (i.e. a nullptr) if
+      /// they have not been used yet.
+      unsigned GetNSlots() const
       {
-         const auto imtPoolSize = ROOT::GetThreadPoolSize();
-         if (!imtPoolSize) {
-            Warning("TThreadedObject()",
-               "Use without IMT is deprecated, either enable IMT first, or use constructor overload taking TNumSlots!");
-         }
-         auto nslots = std::max((unsigned)fgMaxSlots, imtPoolSize);
-         Create(nslots, args...);
+         std::lock_guard<ROOT::TSpinMutex> lg(fSpinMutex);
+         return fObjPointers.size();
       }
 
-      /// Access a particular processing slot. This
-      /// method is *thread-unsafe*: it cannot be invoked from two different
-      /// threads with the same argument.
+      /// Access a particular processing slot.
+      ///
+      /// This method is thread-safe as long as concurrent calls request different slots (i.e. pass a different
+      /// argument) and no thread accesses slot `i` via the arrow operator, so mixing usage of GetAtSlot
+      /// with usage of the arrow operator can be dangerous.
       std::shared_ptr<T> GetAtSlot(unsigned i)
       {
-         if ( i >= fObjPointers.size()) {
-            Warning("TThreadedObject::GetAtSlot", "Maximum number of slots reached.");
+         std::size_t nAvailableSlots;
+         {
+            // fObjPointers can grow due to a concurrent operation on this TThreadedObject, need to lock
+            std::lock_guard<ROOT::TSpinMutex> lg(fSpinMutex);
+            nAvailableSlots = fObjPointers.size();
+         }
+
+         if (i >= nAvailableSlots) {
+            Warning("TThreadedObject::GetAtSlot", "This slot does not exist.");
```

Comment:
So then in the end this doesn't grow the slots? Should it not?

Reply:
Good point.

With the patch in its current form, `TThreadedObject` only grows when access is performed with `Get` or `->` (which make up the simple, sane interface).

When a user asks for a specific slot number through `GetAtSlot`, the problem is that when user requires slot `i`, I would have to create all slots from `nSlots` to `i` included. Or change TThreadedObject's internals a bit more to allow "holes". The other problem is that I am not sure what the expected kind of usage for `GetAtSlot` is.

These are the two usage patterns that I foresaw for `TThreadedObject`:

1. user creates `TThreadedObject(model)` and then has however many threads concurrently accessing its contents via the arrow operator. Each thread gets its own copy of `model`, and new slots are appended as needed (this solves our problem with `TTreeProcessorMT`)
2. as a performance optimization, user can create `TThreadedObject(TNumSlots{4}, model)` and then set the slot contents via `SetAtSlot` or modify the contents via `GetAtSlot`. This is, of course, if they are sure that they won't use more than `TNumSlots{4}` slots _or_ they let threads access slots concurrently via `Get` or `->`, which grows the slots as needed, but creates new slots copying the model, does not let users choose their initial content. We don't have this usage pattern in ROOT afaik

P.S.
I broke a use-case: user creates `TThreadedObject(model)` and then accesses `GetAtSlot(i)` with `i <= fgMaxSlots`: before, `GetAtSlot` would lazily populate the slot number requested with a copy of `model`, now we just return a `nullptr`. Fixing it now.

Reply:
Forgot to ask questions:
* do you think `GetAtSlot(i)` should grow the number of (empty, nullptr) slots to "at least i"?
* how is `GetAtSlot` supposed to be used?

Reply:
Agreed with Axel that it's fine if `GetAtSlot(i)` throws if `i > currentNumberOfSlots`.